pkg/larker/larker.go

```diff
@@ -112,3 +124,98 @@ func (larker *Larker) Main(ctx context.Context, source string) (string, error) {
 
 	return formattedYaml, nil
 }
+
+func (larker *Larker) Hook(
+	ctx context.Context,
+	source string,
+	name string,
+	arguments []interface{},
+) (*HookResult, error) {
+	if name == "" {
+		return nil, fmt.Errorf("%w: empty hook name specified", ErrSanity)
+	}
+
+	if len(arguments) != 1 {
+		return nil, fmt.Errorf("%w: hook takes exactly 1 argument, got %d", ErrSanity, len(arguments))
+	}
+
+	outputLogsBuffer := &bytes.Buffer{}
+
+	thread := &starlark.Thread{
+		Load: loader.NewLoader(ctx, larker.fs, larker.env).LoadFunc(),
+		Print: func(thread *starlark.Thread, msg string) {
+			_, _ = fmt.Fprintln(outputLogsBuffer, msg)
+		},
+	}
+
+	resCh := make(chan *HookResult)
+	errCh := make(chan error)
+
+	go func() {
+		// Execute the source code for the hook to be visible
+		globals, err := starlark.ExecFile(thread, ".cirrus.star", source, nil)
+		if err != nil {
+			errCh <- fmt.Errorf("%w: %v", ErrLoadFailed, err)
+			return
+		}
+
+		// Retrieve hook
+		hook, ok := globals[name]
+		if !ok {
+			errCh <- fmt.Errorf("%w: %s() not found", ErrHookFailed, name)
+			return
+		}
+
+		// Ensure that hook is a function
+		if _, ok := hook.(*starlark.Function); !ok {
+			errCh <- fmt.Errorf("%w: %s is not a function", ErrHookFailed, name)
+			return
+		}
+
+		hookArgument, err := interfaceAsStarlarkValue(arguments[0])
+		if err != nil {
+			errCh <- fmt.Errorf("%w: hook's ctx argument should be JSON-compatible: %v", ErrHookFailed, err)
+			return
+		}
+
+		// Run hook and measure time spent
+		//
+		// We could've used unix.Getrusage() here instead, however:
+		// * it's not clear if we even need such level of precision at the moment
+		// * precise time measurement requires:
+		//   * usage of the Linux-specific RUSAGE_THREAD flag
+		//   * guarding starlark.Call() with runtime.LockOSThread()/runtime.UnlockOSThread()
+		hookStartTime := time.Now()
+
+		hookResult, err := starlark.Call(thread, hook, starlark.Tuple{hookArgument}, nil)
+		if err != nil {
+			errCh <- fmt.Errorf("%w: %v", ErrExecFailed, err)
+			return
+		}
+
+		durationNanos := time.Since(hookStartTime).Nanoseconds()
+
+		// Convert Starlark-style value to interface{}-style value
+		hookResultStarlark, err := starlarkValueAsInterface(hookResult)
+		if err != nil {
+			errCh <- err
+			return
+		}
+
+		// All good
+		resCh <- &HookResult{
+			OutputLogs:    outputLogsBuffer.Bytes(),
+			DurationNanos: durationNanos,
+			Result:        hookResultStarlark,
+		}
+	}()
+
+	select {
+	case hookResult := <-resCh:
+		return hookResult, nil
+	case err := <-errCh:
+		return &HookResult{ErrorMessage: err.Error()}, err
+	case <-ctx.Done():
```

Comment:
Should we do something about `thread` here? Like calling `thread.Cancel("canceled")`

Reply:
Good catch, fixed in https://github.com/cirruslabs/cirrus-cli/pull/303/commits/2711b422c22e4bc2ca140c06f4665b59c2df98c1.